### PR TITLE
Remove routing actor.

### PR DESCRIPTION
**Pull request by pompon0**
Replaced actix routing Actor with synchronous demultiplexed routing table computation. This makes code less asynchronous, so more predictable, so less prone to make tests flaky. Routing table computation still has a dedicated thread, because moving CPU heavy stuff to rayon would require even more invasive refactor. I had to move some code around between the files and that's where the PR size comes from.

I've also removed the graph-related benchmark, because nobody is using them anyway (so they are a deadcode in a way) and they weren't compatible with this refactor. I can write new benchmarks if needed.

### CODE REVIEW

File: chain/network/src/routing/graph/mod.rs [hunk 1]
```diff
@@ -0,0 +1,315 @@
...
+    }
+
+    /// Removes an edge by key. O(1).
+    fn remove_edge(&mut self, key: &EdgeKey) {
+        if self.edges.remove(key).is_some() {
+            self.graph.remove_edge(&key.0, &key.1);
+        }
+    }
+
+    fn remove_adjacent_edges(&mut self, peers: &HashSet<PeerId>) -> Vec<Edge> {
```
Comment (mm-near) [inline chain/network/src/routing/graph/mod.rs:85]: add comment

File: chain/network/src/routing/graph/mod.rs [hunk 2]
```diff
@@ -0,0 +1,315 @@
...
+        }
+    }
+
+    /// Prunes peers unreachable since <unreachable_since> (and their adjacent edges)
+    /// from the in-mem graph and stores them in DB.
+    fn prune_unreachable_peers(&mut self, unreachable_since: time::Instant) {
+        // Select peers to prune.
+        let mut peers = HashSet::new();
+        for k in self.edges.keys() {
+            for peer_id in [&k.0, &k.1] {
```
Comment (mm-near) [inline chain/network/src/routing/graph/mod.rs:156]: seems that we do a lot of unnecessary work here - as we might be looking at the same peer many times.

maybe add a TODO + a tracking bug to fix it.

(might be a good 'first_time_issue' for a new contributor)

File: chain/network/src/routing/graph/mod.rs [hunk 3]
```diff
@@ -0,0 +1,315 @@
...
+        // Prune edges from graph.
+        let edges = self.remove_adjacent_edges(&peers);
+
+        // Store the pruned data in DB.
+        if let Err(e) = self.store.push_component(&peers, &edges) {
+            tracing::warn!("self.store.push_component(): {}", e);
+        }
+    }
+
+    /// 1. Adds edges to the graph.
```
Comment (mm-near) [inline chain/network/src/routing/graph/mod.rs:185]: these edges are validated, right?

File: chain/network/src/routing/graph/mod.rs [hunk 4]
```diff
@@ -0,0 +1,315 @@
...
+        let now = clock.now_utc();
+        for edge in &edges {
+            let key = edge.key();
+            self.load_component(now, key.0.clone());
+            self.load_component(now, key.1.clone());
+        }
+        edges.retain(|e| self.update_edge(now, e.clone()));
+        // Update metrics after edge update
+        metrics::EDGE_UPDATES.inc_by(total as u64);
+        metrics::EDGE_ACTIVE.set(self.graph.total_active_edges() as i64);
```
Comment (mm-near) [inline chain/network/src/routing/graph/mod.rs:211]: nit: shouldn't we update them after the prune_old_edges/prune unreachable peers call?

File: chain/network/src/peer_manager/testonly.rs [hunk 5]
```diff
@@ -7,7 +7,7 @@ use crate::network_protocol::{
 use crate::peer;
 use crate::peer::peer_actor::ClosingReason;
 use crate::peer_manager::network_state::NetworkState;
-use crate::peer_manager::peer_manager_actor;
+//use crate::peer_manager::peer_manager_actor;
```
Comment (mm-near) [inline chain/network/src/peer_manager/testonly.rs:10]: remove?

File: chain/network/src/broadcast/mod.rs [hunk 6]
```diff
@@ -58,13 +58,17 @@ impl<T: Clone + Send> Receiver<T> {
...
     pub async fn recv(&mut self) -> T {
         self.next += 1;
-        let l = self.channel.stream.read().unwrap();
-        let n = self.channel.notify.notified();
-        let v = if l.len() > self.next - 1 { Some(l[self.next - 1].clone()) } else { None };
-        if let Some(v) = v {
-            return v;
-        }
-        drop(l);
+        let n = {
```
Comment (mm-near) [inline chain/network/src/broadcast/mod.rs:61]: please use more descriptive variable names ;-)
and add more comments on what this code does.

**Review by mm-near — APPROVED** [review]

File: chain/network/src/config.rs [hunk 7]
```diff
@@ -320,7 +320,7 @@ impl NetworkConfig {
             inbound_disabled: false,
             archive: false,
             accounts_data_broadcast_rate_limit: demux::RateLimit { qps: 100., burst: 1000000 },
-            routing_table_update_rate_limit: demux::RateLimit { qps: 100., burst: 1000000 },
+            routing_table_update_rate_limit: demux::RateLimit { qps: 10., burst: 1 },
```
[the remark this answers is not in the document]
Comment (pompon0) [reply chain/network/src/config.rs:323]: yes, they improve running time of the tests, which were affected by the changes.

File: chain/network/src/peer_manager/network_state/routing.rs [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+                    }
+                }
+                // Broadcast new edges to all other peers.
+                this.broadcast_routing_table_update(RoutingTableUpdate::from_edges(edges));
+                results
+            })
+            .await;
+        match ok {
+            true => Ok(()),
+            false => Err(ReasonForBan::InvalidEdge),
```
[the remark this answers is not in the document]
Comment (pompon0) [reply chain/network/src/peer_manager/network_state/routing.rs:96]: added a comment

File: chain/network/src/peer_manager/network_state/routing.rs [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+            return;
+        }
+        rtu.edges = Edge::deduplicate(rtu.edges);
+        let msg = Arc::new(PeerMessage::SyncRoutingTable(rtu));
+        for conn in self.tier2.load().ready.values() {
+            conn.send_message(msg.clone());
+        }
+    }
+
+    pub async fn add_accounts(self: &Arc<NetworkState>, accounts: Vec<AnnounceAccount>) {
```
[the remark this answers is not in the document]
Comment (pompon0) [reply chain/network/src/peer_manager/network_state/routing.rs:24]: done

File: chain/network/src/peer_manager/testonly.rs [hunk 10]
```diff
@@ -341,27 +338,20 @@ impl ActorHandler {
...
             let got =
-                self.with_state(|s| async move { s.routing_table_view.info().next_hops }).await;
+                self.with_state(|s| async move { s.graph.routing_table.info().next_hops }).await;
             if test_utils::expected_routing_tables(&got, want) {
                 return;
             }
             events
                 .recv_until(|ev| match ev {
-                    Event::PeerManager(PME::RoutingTableUpdate { .. }) => Some(()),
+                    // Event::PeerManager(PME::RoutingTableUpdate { .. }) => Some(()),
```
[the remark this answers is not in the document]
Comment (pompon0) [reply chain/network/src/peer_manager/testonly.rs:351]: done

File: chain/network/src/routing/graph/mod.rs (hunk 1 above)
Comment (pompon0) [reply chain/network/src/routing/graph/mod.rs:85]: dome

File: chain/network/src/routing/graph/mod.rs (hunk 2 above)
Comment (pompon0) [reply chain/network/src/routing/graph/mod.rs:156]: well, I'm pretty sure it is not worth optimizing:
- we go through all the edges multiple times anyway (bfs, pruning old edges, finding local edges, etc)
- we write to DB in addition to that (unless we don't find anything to prune)
- we first find all the unreachable peers and only then scan for adjacent edges, so it is not like we do a lot of work per peer_id.
- we should kill with fire the graph anyway, so it feels like a potentially wasted effort.

If you still think it is worth creating an issue for a new contributor, please let me know what kind of optimization you would like to see here.

File: chain/network/src/routing/graph/mod.rs (hunk 3 above)
Comment (pompon0) [reply chain/network/src/routing/graph/mod.rs:185]: they are validated in NetworkState::add_edges. Here they are just accepted blindly.

Comment (pompon0) [reply chain/network/src/routing/graph/mod.rs:185]: updated the comment

File: chain/network/src/routing/graph/mod.rs (hunk 4 above)
Comment (pompon0) [reply chain/network/src/routing/graph/mod.rs:211]: makes sense, moved all metric updating to the bottom of the function

File: chain/network/src/peer_manager/testonly.rs (hunk 5 above)
Comment (pompon0) [reply chain/network/src/peer_manager/testonly.rs:10]: done

**Pull request merged**